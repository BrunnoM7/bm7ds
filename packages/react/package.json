{
  "name": "@bm7ds/react",
  "version": "0.2.0",
  "main": "lib/index.js",
  "license": "MIT",
  "devDependencies": {
    "@babel/core": "^7.17.5",
    "@babel/preset-env": "^7.16.11",
    "@babel/preset-react": "^7.16.7",
    "@babel/preset-typescript": "^7.16.7",
    "@storybook/addon-a11y": "^6.4.19",
    "@storybook/addon-actions": "^6.4.19",
    "@storybook/addon-essentials": "^6.4.19",
    "@storybook/addon-interactions": "^6.4.19",
    "@storybook/addon-knobs": "^6.4.0",
    "@storybook/addon-links": "^6.4.19",
    "@storybook/addon-storysource": "^6.4.19",
    "@storybook/preset-scss": "^1.0.3",
    "@storybook/react": "^6.4.19",
    "@storybook/testing-library": "^0.0.9",
    "@testing-library/jest-dom": "^5.16.2",
    "@testing-library/react": "^12.1.2",
    "@types/jest": "^27.4.0",
    "@types/react": "^17.0.38",
    "babel-loader": "^8.2.3",
    "chromatic": "^6.4.3",
    "css-loader": "^6.6.0",
    "jest": "^27.5.1",
    "react": "^17.0.2",
    "rollup": "^2.64.0",
    "rollup-plugin-typescript2": "^0.31.1",
    "sass": "^1.49.8",
    "sass-loader": "^12.6.0",
    "style-loader": "^3.3.1",
    "typescript": "^4.5.4"
  },
  "scripts": {
    "build": "rollup -c",
    "dev": "yarn build --watch",
    "test": "jest --verbose",
    "test:watch": "yarn test --watch",
    "storybook": "start-storybook -p 6006",
    "storybook:build": "yarn build-storybook",
    "build-storybook": "build-storybook",
    "chromatic": "npx chromatic --project-token=2c7e9b7ebb30"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "5bbff22369ef4fc0948687dea4b9ece07a22d1c4"
}
