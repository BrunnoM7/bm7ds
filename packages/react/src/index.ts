import Button from "./atoms/Button"
import Color from "./atoms/Color"
import Image from "./atoms/Image"
import Text from "./atoms/Text"


export {
  Button,
  Color,
  Image,
  Text
}