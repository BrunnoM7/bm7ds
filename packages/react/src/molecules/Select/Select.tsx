import React, { useState, useRef, useEffect, KeyboardEventHandler, createRef } from 'react';
import Text from '../../atoms/Text';

const KEYS = {
  ENTER: "Enter",
  SPACE_BAR: " ",
  DOWN_ARROW: "ArrowDown",
  UP_ARROW: "ArrowUp",
  ESC: "Escape"
}

interface SelectOption {
  label: string
  value: string
}

interface RenderOptionProps {
  isSelected: boolean
  option: SelectOption
  getOptionRecommendedProps: (overrideProps?: Object) => Object
}

interface SelectProps {
  onOptionSelected?: (option: SelectOption, optionIndex: number) => void
  options?: SelectOption[]
  label?: string
  renderOption?: (props: RenderOptionProps) => React.ReactNode
}

const getNextOptionIndex = (currentIndex: number|null, options: Array<SelectOption>) => {
  if (currentIndex === null) {
    return 0;
  }
  if (currentIndex === options.length - 1) {
    return 0;
  }
  return currentIndex + 1;
}

const getPreviousOptionIndex = (currentIndex: number|null, options: Array<SelectOption>) => {
  if (currentIndex === null) {
    return 0;
  }
  if (currentIndex === 0) {
    return options.length - 1;
  }
  return currentIndex - 1;
}

const Select: React.FunctionComponent<SelectProps> = ({ options = [], label = 'Select an option...', onOptionSelected: handler, renderOption }) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [selectedIndex, setSelectedIndex] = useState<number|null>(null);
  const [highlightedIndex, setHighlightedIndex] = useState<number|null>(null);
  const [overlayTop, setOverlayTop] = useState<number>(0);
  const [optionRefs, setOptionRefs] = useState<React.RefObject<HTMLLIElement>[]>([])
  const labelRef = useRef<HTMLButtonElement>(null);

  useEffect(() => {
    setOverlayTop((labelRef.current?.offsetHeight || 0) + 10)
  }, [labelRef.current?.offsetHeight]);

  useEffect(() => {
    setOptionRefs(options.map(_ => createRef<HTMLLIElement>()))
  }, [options.length]);
  
  useEffect(() => {
    if(highlightedIndex !== null && isOpen) {
      const ref = optionRefs[highlightedIndex];

      if(ref && ref.current) {
        ref.current.focus();
      }
    }
  }, [isOpen, highlightedIndex])

  const onOptionSelected = (option: SelectOption, optionIndex: number) => {
    setIsOpen(!isOpen);
    if(handler) {
      handler(option, optionIndex)
    }
    setSelectedIndex(optionIndex)
  }

  const onLabelClick = () => {
    setIsOpen(!isOpen);
  }

  const highlightOption = (optionIndex: number|null) => {
    setHighlightedIndex(optionIndex);
  }

  const onButtonKeyDown: KeyboardEventHandler = (event) => {
    event.preventDefault();

    if([KEYS.ENTER, KEYS.SPACE_BAR, KEYS.DOWN_ARROW].includes(event.key)) {
      setIsOpen(true);

      // set focus on the list item
      highlightOption(0);
    }
  }

  const onOptionKeyDown: KeyboardEventHandler = (event) => {
    if(event.key === KEYS.ESC) {
      setIsOpen(false);

      return
    }

    if(event.key === KEYS.DOWN_ARROW) {
      highlightOption(getNextOptionIndex(highlightedIndex, options));
    }
    if(event.key === KEYS.UP_ARROW) {
      highlightOption(getPreviousOptionIndex(highlightedIndex, options));
    }
    if(event.key === KEYS.ENTER) {
      onOptionSelected(options[highlightedIndex!], highlightedIndex!);
    }
  }

  let selectedOption = null;
  if(selectedIndex !== null) {
    selectedOption = options[selectedIndex];
  }

  return <div className='bmds-select'>
    <button 
      className='bmds-select__label' 
      ref={labelRef} 
      aria-haspopup={true} 
      aria-expanded={isOpen? true : undefined} 
      aria-controls='bmds-select-list'
      onClick={() => onLabelClick()}
      onKeyDown={onButtonKeyDown}
      tabIndex={0}
      data-testid="BmdsSelectButton"
    >
      <Text>{selectedOption === null ? label : selectedOption.label }</Text>
      
      {/* 
        THIS ICON IS FROM HEROICONS LIBRARY PASTED DIRECTLY INTO THIS CODE 
        A BETTER ICON IMPLEMENTATION IS NEEDED
      */}
      <svg className={`bmds-select__caret ${ isOpen ? 'bmds-select__caret--open' : 'bmds-select__caret--closed' }`} xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor">
        <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M19 9l-7 7-7-7" />
      </svg>
    </button>

    {
      <ul 
        role='menu' 
        id='bmds-select-list' 
        style={{ top: overlayTop }} 
        className={`bmds-select__overlay ${ isOpen ? 'bmds-select__overlay--open' : '' }`}

      >
        {options.map((option, optionIndex) => {
          const isSelected = selectedIndex === optionIndex;
          const isHighlighted = highlightedIndex === optionIndex;

          const ref = optionRefs[optionIndex];

          const renderOptionProps = {
            option,
            isSelected,
            getOptionRecommendedProps: (overrideProps = {}) => {
              return {
                ref,
                role: 'menuitemradio',
                'aria-label': option.label,
                'aria-checked': isSelected ? true : undefined,
                tabIndex: isHighlighted ? -1 : 0,
                className: `bmds-select__option
                  ${isSelected ? 'bmds-select__option--selected' : ''}
                  ${isHighlighted ? 'bmds-select__option--highlighted' : ''}
                `,
                key: option.value,
                onClick: () => onOptionSelected(option, optionIndex),
                onMouseEnter: () => highlightOption(optionIndex),
                onMouseLeave: () => highlightOption(null),
                onKeyDown: onOptionKeyDown,
                ...overrideProps
              }
            }
          }

          if (renderOption) {
            return renderOption(renderOptionProps)
          }
          return (
            <li 
              {...renderOptionProps.getOptionRecommendedProps()}
            >
              <Text>
                {option.label}
              </Text>

              {/* 
                THIS ICON IS FROM HEROICONS LIBRARY PASTED DIRECTLY INTO THIS CODE 
                A BETTER ICON IMPLEMENTATION IS NEEDED
              */}

              {isSelected ? (
                <svg className={`bmds-select__caret`} xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                  <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M5 13l4 4L19 7" />
                </svg>
              ) : null}
            </li>)
        })}
      </ul>
    }
  </div>;
};

export default Select;
