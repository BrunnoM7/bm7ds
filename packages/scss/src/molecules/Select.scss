@import "foundation/all";

.bmds-select {
  display: flex;
  flex-direction: column;

  position: relative;

  &__label {
    display: flex;
    align-items: center;
    justify-content: space-between;

    @include padding("xs");
    border: 1px solid $form-border-color;
    background-color: $white;

    font-family: $body-font-family;

    cursor: pointer;
  }

  &__overlay {
    @include padding("none");
    @include margin("none");

    border: 1px solid $form-border-color;
    background-color: $form-bg-color;
    z-index: 99;

    list-style-type: none;

    box-shadow: map-get($shadows, "base");

    position: absolute;
    width: 100%;
  }

  &__option {
    display: flex;
    align-items: center;
    justify-content: space-between;
    width: 100%;
    cursor: pointer;
    @include padding("xs");

    box-sizing: border-box;
  }

  &__option:hover {
    background-color: $form-bg-color-hover;
  }

  &__option--selected {
    background-color: $form-bg-option-selected;
    color: $form-color-option-selected;
  }

  &__option--selected:hover {
    background-color: $form-bg-option-selected-hover;
  }

  &__caret {
    transition: all 250ms ease;
    width: $fs-base;
    height: $fs-base;
  }

  &__caret--open {
    transform: rotate(180deg);
  }

  &__caret--closed {
    transform: rotate(0deg);
  }
}
