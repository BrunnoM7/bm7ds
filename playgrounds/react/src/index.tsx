import React from "react";
import ReactDOM from "react-dom";

import App from './App';
import '@bmds/scss/lib/global.css'

ReactDOM.render(
  <App />,
  document.querySelector('#root')
);