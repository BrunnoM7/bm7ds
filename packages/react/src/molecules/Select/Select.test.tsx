import React from "react";
import Select from "./Select";

import { render, fireEvent } from '@testing-library/react';

const options = [
  {
    label: 'Strict Black',
    value: "strict-black"
  },
  {
    label: 'Black Lagoon',
    value: "black-lagoon"
  },
  {
    label: 'Sirius Black',
    value: "sirius-black"
  },
]

test('renders all options passed to it', () => {
  const { getAllByRole, getByTestId } = render(<Select options={options} />);

  fireEvent.click(getByTestId('BmdsSelectButton'));

  expect(getAllByRole('menuitemradio')).toHaveLength(options.length);
})

test('can customize select label', () => {
  const { getByText } = render(<Select options={options} label='CUSTOM LABEL'/>);

  expect(getByText('CUSTOM LABEL')).toBeInTheDocument()
})

test('renders options using custom renderOption method if passed as prop', () => {
  const { getAllByTestId, getByTestId } = render(<Select options={options} renderOption={({ option, getOptionRecommendedProps }) => {
    return <li data-testid='CustomRenderOption' { ...getOptionRecommendedProps() }>{option.label}</li>
  }} />)

  fireEvent.click(getByTestId('BmdsSelectButton'));

  expect(getAllByTestId('CustomRenderOption')).toHaveLength(options.length)
})

test('calls the onOptionSelected prop with the selected option and its index if passed', () => {
  const onOptionSelected = jest.fn();
  const { getAllByRole, getByTestId } = render(<Select options={options} onOptionSelected={onOptionSelected}/>);
  
  fireEvent.click(getByTestId('BmdsSelectButton'));
  fireEvent.click(getAllByRole('menuitemradio')[0]);

  expect(onOptionSelected).toHaveBeenCalledWith(options[0], 0);
})

test('the button label changes to the selected option label', () => {
  const { getAllByRole, getByTestId } = render(<Select options={options} />);
  
  fireEvent.click(getByTestId('BmdsSelectButton'));
  fireEvent.click(getAllByRole('menuitemradio')[0]);

  expect(getByTestId('BmdsSelectButton')).toHaveTextContent(options[0].label);
  
})

test('snapshot of the selected option state', () => {
  const { getAllByRole, getByTestId, asFragment } = render(<Select options={options} />);
  
  fireEvent.click(getByTestId('BmdsSelectButton'));
  fireEvent.click(getAllByRole('menuitemradio')[0]);

  expect(asFragment()).toMatchSnapshot();
})

test('snapshot of the base state', () => {
  const { asFragment } = render(<Select options={options} />);

  expect(asFragment()).toMatchSnapshot();
})

test('snapshot of the options menu open state', () => {
  const { getByTestId, asFragment } = render(<Select options={options} />);
  
  fireEvent.click(getByTestId('BmdsSelectButton'));

  expect(asFragment()).toMatchSnapshot();
})