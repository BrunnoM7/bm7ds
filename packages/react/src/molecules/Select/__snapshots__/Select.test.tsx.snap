// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="bmds-select"
  >
    <button
      aria-controls="bmds-select-list"
      aria-haspopup="true"
      class="bmds-select__label"
      data-testid="BmdsSelectButton"
      tabindex="0"
    >
      <p
        class="bmds-text bmds-text-md"
      >
        Select an option...
      </p>
      <svg
        class="bmds-select__caret bmds-select__caret--closed"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the options menu open state 1`] = `
<DocumentFragment>
  <div
    class="bmds-select"
  >
    <button
      aria-controls="bmds-select-list"
      aria-expanded="true"
      aria-haspopup="true"
      class="bmds-select__label"
      data-testid="BmdsSelectButton"
      tabindex="0"
    >
      <p
        class="bmds-text bmds-text-md"
      >
        Select an option...
      </p>
      <svg
        class="bmds-select__caret bmds-select__caret--open"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
    <ul
      class="bmds-select__overlay"
      id="bmds-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Strict Black"
        class="bmds-select__option
                  
                  
                "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="bmds-text bmds-text-md"
        >
          Strict Black
        </p>
      </li>
      <li
        aria-label="Black Lagoon"
        class="bmds-select__option
                  
                  
                "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="bmds-text bmds-text-md"
        >
          Black Lagoon
        </p>
      </li>
      <li
        aria-label="Sirius Black"
        class="bmds-select__option
                  
                  
                "
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="bmds-text bmds-text-md"
        >
          Sirius Black
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`snapshot of the selected option state 1`] = `
<DocumentFragment>
  <div
    class="bmds-select"
  >
    <button
      aria-controls="bmds-select-list"
      aria-haspopup="true"
      class="bmds-select__label"
      data-testid="BmdsSelectButton"
      tabindex="0"
    >
      <p
        class="bmds-text bmds-text-md"
      >
        Strict Black
      </p>
      <svg
        class="bmds-select__caret bmds-select__caret--closed"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
